### PR TITLE
Fix cell complex methods

### PR DESCRIPTION
**Pull request by mhajij**
the following methods were added

1- get_linegraph : the simplest graph that can be attached to a cell complex
2- all_cell_to_node_codjacnecy_matrix 
3-node_to_all_cell_adjacnecy_matrix

based on the above the following methods were corrected

1- s_connected_components
2- s_component_subcomplexes
3- connected_components
4- connected_component_subcomplexes
5- node_diameters
6- cell_diameters
7- diameter
8- cell_diameter
9- distance
10- cell_distance

### CODE REVIEW

File: toponetx/classes/cell_complex.py [hunk 1]
```diff
@@ -130,10 +130,10 @@ class CellComplex(Complex):
...
     """
 
-    def __init__(
-        self, cells=None, name: str = "", regular: bool = True, **kwargs
-    ) -> None:
-        super().__init__(name, **kwargs)
+    def __init__(self, cells=None, name: str = "", regular=True, **kwargs) -> None:
+        super().__init__(**kwargs)
+
+        self.name = name
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:136]:
```suggestion
        super().__init__(name, **kwargs)
```
(No need to set `self.name` if passing the name to the superclass)

File: toponetx/classes/cell_complex.py [hunk 2]
```diff
@@ -396,14 +396,14 @@ def _remove_equivalent_cells(self) -> None:
                         else:
                             self._delete_cell(c, k)
 
-    def degree(self, node: Hashable, rank: int = 1) -> int:
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:399]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 3]
```diff
@@ -396,14 +396,14 @@ def _remove_equivalent_cells(self) -> None:
...
+    def degree(self, node: Hashable, rank=1) -> int:
         """Compute the number of cells of certain rank that contain node.
 
         Parameters
         ----------
         node : hashable
             Identifier for the node.
-        rank : int, default=1
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:406]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 4]
```diff
@@ -542,13 +542,14 @@ def neighbors(self, node):
 
         return self._G[node]
 
-    def cell_neighbors(self, cell, s: int = 1):
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:545]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 5]
```diff
@@ -542,13 +542,14 @@ def neighbors(self, node):
...
+    def cell_neighbors(self, cell, s=1):
         """Cells in cell complex which share s nodes(s) with cells.
 
         Parameters
         ----------
         cell : Cell or Iterable representing a acell
-        s : int, default=1
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:551]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 6]
```diff
@@ -558,7 +559,7 @@ def cell_neighbors(self, cell, s: int = 1):
         """
         raise NotImplementedError()
 
-    def remove_node(self, node: Hashable) -> None:
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:561]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 7]
```diff
@@ -600,11 +601,11 @@ def remove_nodes(self, node_set: Iterable[Hashable]) -> None:
         for node in node_set:
             self.remove_node(node)
 
-    def add_node(self, node: Hashable, **attr) -> None:
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:603]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 8]
```diff
@@ -600,11 +601,11 @@ def remove_nodes(self, node_set: Iterable[Hashable]) -> None:
...
+    def add_node(self, node: Hashable, **attr):
         """Add a single node to cell complex."""
         self._G.add_node(node, **attr)
 
-    def _add_nodes_from(self, nodes: Iterable[Hashable]) -> None:
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:607]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 9]
```diff
@@ -614,7 +615,7 @@ def _add_nodes_from(self, nodes: Iterable[Hashable]) -> None:
         for node in nodes:
             self.add_node(node)
 
-    def add_edge(self, u_of_edge: Hashable, v_of_edge: Hashable, **attr) -> None:
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:617]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 10]
```diff
@@ -625,7 +626,7 @@ def add_edge(self, u_of_edge: Hashable, v_of_edge: Hashable, **attr) -> None:
         """
         self._G.add_edge(u_of_edge, v_of_edge, **attr)
 
-    def add_edges_from(self, ebunch_to_add: Iterable[tuple], **attr) -> None:
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:628]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 11]
```diff
@@ -640,7 +641,7 @@ def add_cell(
         self,
         cell: tuple | list | Cell,
         rank: int | None = None,
-        check_skeleton: bool = False,
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:643]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 12]
```diff
@@ -731,9 +732,9 @@ def add_cells_from(
         self,
         cell_set: Iterable[tuple | list | Cell],
         rank: int | None = None,
-        check_skeleton: bool = False,
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:734]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 13]
```diff
@@ -1140,7 +1141,7 @@ def get_cell_attributes(self, name: str, rank: int) -> dict[Hashable | tuple, An
             return d
         raise TopoNetXError(f"Rank must be 0, 1 or 2, got {rank}")
 
-    def set_cell_data(self, cell, rank, attr_name: str, attr_value):
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:1143]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 14]
```diff
@@ -1183,7 +1184,7 @@ def set_cell_data(self, cell, rank, attr_name: str, attr_value):
         else:
             raise ValueError(f"Invalid rank: {rank}. Rank must be 0, 1, or 2.")
 
-    def get_cell_data(self, cell, rank, attr_name: str | None = None):
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:1186]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 15]
```diff
@@ -1241,7 +1242,7 @@ def get_cell_data(self, cell, rank, attr_name: str | None = None):
             else:
                 raise KeyError(f"Cell '{cell}' is not present in the complex.")
 
-    def remove_equivalent_cells(self) -> None:
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:1244]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 16]
```diff
@@ -1328,6 +1329,132 @@ def is_insertable_cycle(
...
+    ) -> scipy.sparse.csc_matrix | tuple[dict, dict, scipy.sparse.csc_matrix]:
+        """Nodes/cells incidence matrix for the indexed by nodes X cells.
+
+        Parameters
+        ----------
+        weight : bool, default=False
+            If False all nonzero entries are 1.
+            If True and self.static all nonzero entries are filled by
+            self.cells.cell_weight dictionary values.
+        index : boolean, optional, default False
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:1343]:
```suggestion
        index : bool, default=False
```

File: toponetx/classes/cell_complex.py [hunk 17]
```diff
@@ -1328,6 +1329,132 @@ def is_insertable_cycle(
...
+        Two nodes are s-adjacent iff there exists a cell (1 dimensional or 2 dimensional)
+        share contain them.
+
+        Parameters
+        ----------
+        weight : bool, default=False
+            If False all nonzero entries are 1.
+            If True and self.static all nonzero entries are filled by
+            self.cells.cell_weight dictionary values.
+        index : boolean, optional, default False
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:1384]:
```suggestion
        index : bool, default=False
```

File: toponetx/classes/cell_complex.py [hunk 18]
```diff
@@ -1328,6 +1329,132 @@ def is_insertable_cycle(
...
+        Two cells (1 dimensional or 2 dimensional) are s-coadjacent iff
+        they share a vertex
+
+        Parameters
+        ----------
+        weight : bool, default=False
+            If False all nonzero entries are 1.
+            If True and self.static all nonzero entries are filled by
+            self.cells.cell_weight dictionary values.
+        index : boolean, optional, default False
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:1437]:
```suggestion
        index : bool, default=False
```

File: toponetx/classes/cell_complex.py [hunk 19]
```diff
@@ -1483,7 +1610,7 @@ def incidence_matrix(
             raise ValueError(f"Only dimensions 0, 1 and 2 are supported, got {rank}.")
 
     def hodge_laplacian_matrix(
-        self, rank: int, signed: bool = True, weight: bool = False, index: bool = False
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:1486]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 20]
```diff
@@ -1911,7 +2038,7 @@ def to_hypergraph(self):
         HG._add_nodes_from(nodes)
         return HG
 
-    def is_connected(self, s: int = 1, cells: bool = False):
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:1914]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 21]
```diff
@@ -1979,14 +2106,50 @@ def clone(self) -> "CellComplex":
...
+        Two vertices are connected by an s-line-graph edge if the corresponding cell
+        complex edges intersect in at least s cell complex nodes.
+
+        If cells=False, the cell complex nodes will be the vertices of the line graph.
+        Two vertices are connected if the nodes they correspond to share at least s
+        incident cell complex edges.
+
+        Parameters
+        ----------
+        s : int
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:2131]:
```suggestion
        s : int, default=1
```

File: toponetx/classes/cell_complex.py [hunk 22]
```diff
@@ -1979,14 +2106,50 @@ def clone(self) -> "CellComplex":
...
+            The width of the connections.
+        cells : bool, optional, default=True
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:2133]:
```suggestion
        cells : bool, default=True
```

File: toponetx/classes/cell_complex.py [hunk 23]
```diff
@@ -1979,14 +2106,50 @@ def clone(self) -> "CellComplex":
...
+            Determines if cells or nodes will be the vertices in the line graph.
+
+        Returns
+        -------
+        nx.Graph
+            A NetworkX graph representing the s-linegraph of the Cell Complex.
+        """
+        if isinstance(s, None):
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:2141]: Remove this check

File: toponetx/classes/cell_complex.py [hunk 24]
```diff
@@ -1979,14 +2106,50 @@ def clone(self) -> "CellComplex":
...
+                "s must be a positive integer larger than 1, got type of s None."
+            )
+        if cells:
+            M = self.all_cell_to_node_codjacnecy_matrix(s=s)
+        else:
+            M = self.node_to_all_cell_adjacnecy_matrix(s=s)
+
+        return nx.from_scipy_sparse_array(M)
+
+    def s_connected_components(self, s=1, cells=True, return_singletons=False):
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:2152]: Please add the type hints back (for some reason GitHub shows the old line somewhat further above)

File: toponetx/classes/cell_complex.py [hunk 25]
```diff
@@ -2017,27 +2180,39 @@ def s_connected_components(
...
             Iterator returns sets of uids of the cells (or nodes) in the s-cells(node)
             components of CX.
+
+        Example
+        -------
+        >>> CX = CellComplex()
+        >>> CX.add_cell([2,3,4],rank=2)
+        >>> CX.add_cell([5,6,7],rank=2)
+        >>> list(CX.s_connected_components(s=1,cells=False))
+        >>> # [{2, 3, 4, 5, 6, 7}]
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:2190]:
```suggestion
        [{2, 3, 4, 5, 6, 7}]
```

File: toponetx/classes/cell_complex.py [hunk 26]
```diff
@@ -2017,27 +2180,39 @@ def s_connected_components(
...
+        >>> CX.add_cell([4,5],rank=1)
+        >>> list(CX.s_connected_components(s=1,cells=False))
+        >>> # [{2, 3, 4}, {5, 6, 7}]
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:2193]:
```suggestion
        [{2, 3, 4}, {5, 6, 7}]
```

File: toponetx/classes/cell_complex.py [hunk 27]
```diff
@@ -2017,27 +2180,39 @@ def s_connected_components(
...
             for c in nx.connected_components(G):
                 if not return_singletons:
                     if len(c) == 1:
                         continue
-                yield {rowdict[n] for n in c}
+                yield {node_dict[n] for n in c}
 
-    def s_component_subgraphs(
-        self, s: int = 1, cells: bool = True, return_singletons: bool = False
-    ):
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:2040]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 28]
```diff
@@ -2047,40 +2222,36 @@ def s_component_subgraphs(
...
-        See Also
-        --------
-        s_connected_components
-        """
-        return self.s_connected_components(
-            s=s, cells=cells, return_singletons=return_singletons
-        )
-
-    def connected_components(self, cells: bool = False, return_singletons: bool = True):
+    def connected_components(self, cells=False, return_singletons=True):
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:2254]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 29]
```diff
@@ -2090,43 +2261,37 @@ def connected_components(self, cells: bool = False, return_singletons: bool = Tr
...
+        >>> CX.add_cell([2,3,4],rank=2)
+        >>> CX.add_cell([5,6,7],rank=2)
+        >>> list(CX.connected_components(s=1,cells=False))
+        >>> CX.add_cell([4,5],rank=1)
+        >>> list(CX.connected_components(s=1,cells=False))
         """
         return self.s_connected_components(cells=cells, return_singletons=True)
 
-    def connected_component_subgraphs(self, return_singletons: bool = True):
+    def connected_component_subcomplexes(self, return_singletons=True):
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:2276]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 30]
```diff
@@ -2135,9 +2300,19 @@ def node_diameters(self):
...
         list of the diameters of the s-components and
         list of the s-component nodes
+
+        Example
+        -------
+        >>> CX = CellComplex()
+        >>> CX.add_cell([2,3,4],rank=2)
+        >>> CX.add_cell([5,6,7],rank=2)
+        >>> list(CX.node_diameters())
+
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:2310]: Remove empty line

File: toponetx/classes/cell_complex.py [hunk 31]
```diff
@@ -2147,10 +2322,9 @@ def node_diameters(self):
                 temp.add(coldict[e])
             comps.append(temp)
             diams.append(diamc)
-        loc = np.argmax(diams)
-        return diams[loc], diams, comps
+        return diams, comps
 
-    def cell_diameters(self, s: int = 1):
+    def cell_diameters(self, s=1):
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:2327]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 32]
```diff
@@ -2190,7 +2372,7 @@ def diameter(self):
 
         Returns
         -------
-        int
+        diameter : int
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:2375]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 33]
```diff
@@ -2219,7 +2401,7 @@ def cell_diameter(self, s=1):
 
         Return
         ------
-        int
+        cell_diameter : int
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:2404]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 34]
```diff
@@ -2228,18 +2410,18 @@ def cell_diameter(self, s=1):
...
-        A = self.coadjacency_matrix(rank=2)
-        G = nx.from_scipy_sparse_matrix(A)
+        A = self.all_cell_to_node_codjacnecy_matrix()
+        G = nx.from_scipy_sparse_array(A)
         if nx.is_connected(G):
             return nx.diameter(G)
         raise TopoNetXError(f"cell complex is not s-connected. s={s}")
 
-    def distance(self, source, target, s: int = 1) -> int:
+    def distance(self, source, target, s=1):
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:2424]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 35]
```diff
@@ -2253,7 +2435,7 @@ def distance(self, source, target, s: int = 1) -> int:
 
         Returns
         -------
-        int
+        s-walk distance : int
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:2438]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 36]
```diff
@@ -2273,17 +2455,16 @@ def distance(self, source, target, s: int = 1) -> int:
...
         try:
-            path = nx.shortest_path_length(G, rkey[source], rkey[target])
+            path = nx.shortest_path_length(G, rowdict[source], rowdict[target])
             return path
         except Exception:
             warnings.warn(f"No {s}-path between {source} and {target}")
             return np.inf
 
-    def cell_distance(self, source, target, s: int = 1) -> int:
+    def cell_distance(self, source, target, s=1):
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:2467]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 37]
```diff
@@ -2336,8 +2517,11 @@ def from_networkx_graph(self, G):
         Examples
         --------
         >>> CX = CellComplex()
-        >>> CX.add_cells_from([[1, 2, 4], [1, 2, 7]], rank=2)
-        >>> G = Graph([(0, 1), (0, 2), (1, 2)])
+        >>> CX.add_cells_from([[1,2,4],[1,2,7] ],rank=2)
+        >>> G = Graph()
+        >>> G.add_edge(1,0)
+        >>> G.add_edge(2,0)
+        >>> G.add_edge(1,2)
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:2524]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 38]
```diff
@@ -2384,7 +2568,7 @@ def from_trimesh(cls, mesh) -> "CellComplex":
         return CX
 
     @classmethod
-    def load_mesh(cls, file_path, process: bool = False, force=None) -> "CellComplex":
+    def load_mesh(cls, file_path, process=False, force=None) -> "CellComplex":
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:2571]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 39]
```diff
@@ -239,19 +239,19 @@ def is_regular(self) -> bool:
                 return False
         return True
 
-    def __str__(self) -> str:
+    def __str__(self):
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:242]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 40]
```diff
@@ -239,19 +239,19 @@ def is_regular(self) -> bool:
...
         """Return detailed string representation."""
         return f"Cell Complex with {len(self.nodes)} nodes, {len(self.edges)} edges  and {len(self.cells)} 2-cells "
 
     def __repr__(self) -> str:
         """Return string representation."""
         return f"CellComplex(name='{self.name}')"
 
-    def __len__(self) -> int:
+    def __len__(self):
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:250]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 41]
```diff
@@ -239,19 +239,19 @@ def is_regular(self) -> bool:
...
         """Return number of nodes."""
         return len(self.nodes)
 
-    def __iter__(self) -> Iterator:
+    def __iter__(self):
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:254]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 42]
```diff
@@ -261,7 +261,7 @@ def __iter__(self) -> Iterator:
         """
         return iter(self.nodes)
 
-    def __contains__(self, item) -> bool:
+    def __contains__(self, item):
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:264]: Please undo.

File: toponetx/classes/cell_complex.py [hunk 43]
```diff
@@ -364,7 +364,7 @@ def _cell_equivalence_class(self) -> dict[Cell, set[Cell]]:
                         all_inserted_cells.add(j)
         return equiv_classes
 
-    def _remove_equivalent_cells(self) -> None:
+    def _remove_equivalent_cells(self):
```
Comment (ffl096) [inline toponetx/classes/cell_complex.py:367]: Please undo.

**Review by ffl096 — CHANGES_REQUESTED** [review]
Not sure if this was a bad merge or something, but this pull request removes many recent changes to `cell_complex.py`. Please fix :)

**Comment by mhajij** [comment]
> Not sure if this was a bad merge or something, but this pull request removes many recent changes to `cell_complex.py`. Please fix :)

fixed now.

**Review by USFCA-MSDS — APPROVED** [review]

**Review by USFCA-MSDS — APPROVED** [review]

**Pull request merged**